centrifuge/src/main/java/io/github/centrifugal/centrifuge/Subscription.java

```diff
@@ -41,12 +46,38 @@ SubscriptionEventListener getListener() {
         this.listener = listener;
     }
 
+    public long getLastOffset() {
+        return lastOffset;
+    }
+
+    public void setLastOffset(long lastOffset) {
+        this.lastOffset = lastOffset;
+    }
+
+    public String getLastEpoch() {
+        return lastEpoch;
+    }
+
+    public void setLastEpoch(String lastEpoch) {
+        this.lastEpoch = lastEpoch;
+    }
+
     void moveToUnsubscribed() {
         this.state = SubscriptionState.UNSUBSCRIBED;
     }
 
     void moveToSubscribeSuccess(Protocol.SubscribeResult result) {
         this.state = SubscriptionState.SUBSCRIBED;
+        this.lastEpoch = result.getEpoch();
+        this.lastOffset = result.getOffset();
+        this.setRecover(true);
```

Comment:
[There is a field](https://github.com/centrifugal/protocol/blob/master/definitions/client.proto#L134) `recoverable`  - you can rely on it to understand whether subscription can be recovered.

Reply:
> [There is a field](https://github.com/centrifugal/protocol/blob/master/definitions/client.proto#L134) `recoverable` - you can rely on it to understand whether subscription can be recovered (since not all can - i.e. depends on server options for channel).

ok tks u. I'm used it in last commit.